src/storage/src/render/persist_sink.rs

```diff
@@ -131,17 +144,31 @@ pub fn render<G>(
                     // per-timestamp.
                     for (row, ts, diff) in buffer.drain(..) {
                         if write.upper().less_equal(&ts) {
-                            stashed_batches
-                                .entry(ts)
-                                .or_insert_with(|| {
-                                    // TODO: the lower has to be the min because we don't know
-                                    // what the minimum ts of data we will see is. In the future,
-                                    // this lower should be declared in `finish` instead.
-                                    write.builder(100, Antichain::from_elem(Timestamp::minimum()))
-                                })
+                            let builder = stashed_batches.entry(ts).or_insert_with(|| {
+                                BatchBuilderAndCounts {
+                                    builder: write.builder(
+                                        // TODO: the lower has to be the min because we don't know
+                                        // what the minimum ts of data we will see is. In the future,
+                                        // this lower should be declared in `finish` instead.
+                                        100,
+                                        Antichain::from_elem(Timestamp::minimum()),
+                                    ),
+                                    rows: 0,
+                                    errors: 0,
+                                }
+                            });
+
+                            let ok = row.is_ok();
+                            builder
+                                .builder
                                 .add(&SourceData(row), &(), &ts, &diff)
                                 .await
                                 .expect("invalid usage");
+                            if ok {
+                                builder.rows += 1;
+                            } else {
+                                builder.errors += 1;
+                            }
```

Comment:
note that I don't distinguish between +1 and -1 diffs here, im just counting the count of rows/errors we are writing to persist; perhaps for errors we should use a gauge and dec if we see a negative diff?

Reply:
Yeah I think both is better, cause we can always add em to get the total:

`row_inserts`, `error_inserts`, `row_retractions`, and `error_retractions` seem like good names